ToolBox/AlphaVantageDownloader/AlphaVantageDataDownloader.cs

```diff
@@ -0,0 +1,203 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using CsvHelper;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using RestSharp;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+
+namespace QuantConnect.ToolBox.AlphaVantageDownloader
+{
+    /// <summary>
+    /// Alpha Vantage data downloader
+    /// </summary>
+    public class AlphaVantageDataDownloader : IDataDownloader
+    {
+        private readonly IRestClient _avClient;
+
+        /// <summary>
+        /// Construct AlphaVantageDataDownloader with default RestClient
+        /// </summary>
+        /// <param name="apiKey">API key</param>
+        public AlphaVantageDataDownloader(string apiKey) : this(new RestClient(), apiKey)
+        {
+        }
+
+        /// <summary>
+        /// Dependency injection constructor
+        /// </summary>
+        /// <param name="restClient">The <see cref="RestClient"/> to use</param>
+        /// <param name="apiKey">API key</param>
+        public AlphaVantageDataDownloader(IRestClient restClient, string apiKey)
+        {
+            _avClient = restClient;
+            _avClient.BaseUrl = new Uri("https://www.alphavantage.co/");
+            _avClient.Authenticator = new AlphaVantageAuthenticator(apiKey);
+        }
+
+        /// <summary>
+        /// Get data from API
+        /// </summary>
+        /// <param name="symbol">Symbol to download</param>
+        /// <param name="resolution">Resolution to download</param>
+        /// <param name="startUtc">Start time</param>
+        /// <param name="endUtc">End time</param>
+        /// <returns></returns>
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            var request = new RestRequest("query", DataFormat.Json);
+            request.AddParameter("symbol", symbol.Value);
+            request.AddParameter("datatype", "csv");
+
+            IEnumerable<TimeSeries> data = null;
+            switch (resolution)
+            {
+                case Resolution.Minute:
+                case Resolution.Hour:
+                    data = GetIntradayData(request, startUtc, endUtc, resolution);
+                    break;
+                case Resolution.Daily:
+                    data = GetDailyData(request, startUtc, endUtc);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(resolution), $"{resolution} resolution not supported by API.");
+            }
+
+            var period = resolution.ToTimeSpan();
+            return data.Select(d => new TradeBar(d.Time, symbol, d.Open, d.High, d.Low, d.Close, d.Volume, period));
```

Comment:
Hi,
I cloned this pull request on locally, I am running on a Mac OS using mono but I am seeing the following issue when executing ```mono QuantConnect.ToolBox.exe 
    --app=alphavantagedownloader 
    --tickers=SPXL 
    --resolution=Hour 
    --from-date=20210101-00:00:00 \
    --api-key=```
```20210423 11:48:52.933 ERROR:: AlphaVantageDownloaderProgram.AlphaVantageDownloader():  CsvHelper.HeaderValidationException: Header with name 'time' or 'timestamp'[0] was not found.
Header with name 'open'[0] was not found.
Header with name 'high'[0] was not found.
Header with name 'low'[0] was not found.
Header with name 'close'[0] was not found.
Header with name 'volume'[0] was not found.
If you are expecting some headers to be missing and want to ignore this validation, set the configuration HeaderValidated to null. You can also change the functionality to do something else, like logging the issue.

  at CsvHelper.Configuration.ConfigurationFunctions.HeaderValidated (CsvHelper.InvalidHeader[] invalidHeaders, CsvHelper.ReadingContext context) [0x0006d] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader.ValidateHeader (System.Type type) [0x000af] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader.ValidateHeader[T] () [0x00000] in <e0ff1191374f4bc2b71069db98504b37>:0
  at CsvHelper.CsvReader+<GetRecords>d__63`1[T].MoveNext () [0x0006f] in <e0ff1191374f4bc2b71069db98504b37>:0
  at System.Collections.Generic.EnumerableHelpers.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source, System.Int32& length) [0x00032] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Buffer`1[TElement]..ctor (System.Collections.Generic.IEnumerable`1[T] source) [0x00022] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.OrderedEnumerable`1[TElement].ToList () [0x00000] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDataDownloader.GetTimeSeries (RestSharp.RestRequest request) [0x00052] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDataDownloader+<GetIntradayData>d__5.MoveNext () [0x00136] in <8e97fb96193a498bad25e020041ad1a8>:0
  at System.Linq.Enumerable+SelectEnumerableIterator`2[TSource,TResult].MoveNext () [0x00029] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector, System.Collections.Generic.IEqualityComparer`1[T] comparer) [0x000a3] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector) [0x00000] in <ef2d591aaba4433ebe1b6477bc29c58b>:0
  at QuantConnect.ToolBox.LeanDataWriter.WriteDailyOrHour (System.Collections.Generic.IEnumerable`1[T] source) [0x0001d] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.LeanDataWriter.Write (System.Collections.Generic.IEnumerable`1[T] source) [0x00018] in <8e97fb96193a498bad25e020041ad1a8>:0
  at QuantConnect.ToolBox.AlphaVantageDownloader.AlphaVantageDownloaderProgram.AlphaVantageDownloader (System.Collections.Generic.List`1[T] tickers, System.String resolution, System.DateTime fromDate, System.DateTime toDate, System.String apiKey) [0x00100] in <8e97fb96193a498bad25e020041ad1a8>:0```

any ideas how to fix it?

Reply:
@nuaimat It looks like you might be getting rate limited. You can try with a shorter `from-date` and `to-date` window.

Unfortunately, the API always responds with a `200 OK` regardless of status. I'll add some extra logic to look for content other than CSV data and display the message to the user.

Reply:
@jdharmon that's correct, printing the values inside GetTimeSeries shows the rate limit message from the api,  
adding `System.Threading.Thread.Sleep(15*1000);` inside GetIntradayData `foreach (var slice in slices)`
seems to do the trick for me, would love if this was parametrized "--delayBetweenRequests"

Reply:
Added rate limiting.